lib/spack/spack/build_systems/rocm.py

```diff
@@ -145,7 +145,12 @@ class ROCmPackage(PackageBase):
     depends_on("hip +rocm", when="+rocm")
 
     # need amd gpu type for rocm builds
+    compilers_supporting_rocm = ("cce", "rocmcc", "clang", "aocc")
     conflicts("amdgpu_target=none", when="+rocm")
+    # If this variable shadows a property, it overrides it
+    for cmp_name in spack.compilers.supported_compilers():
+        if cmp_name not in compilers_supporting_rocm:
+            conflicts(f"%{cmp_name}", when="+rocm")
```

Comment:
I think this would be a mistake to enforce for _all_ `ROCmPackage`s. This breaks projects that use CMake's `enable_language(HIP)` properly, i.e. use e.g. gcc for host compilation and HIP for device code; exactly like for CUDA. If there are some projects that require using hipcc or similar to compile the whole project, the conflicts could be in those packages instead?

Reply:
There is currently not a single package in Spack using `CMAKE_HIP_COMPILER` to properly separate cxx from hip compilation. I 100% agree that we need to support this better in the future, but currently my understanding is all of these packages are instead lying to Spack about which compiler is used and then just using the hip compiler for all compilation.

Reply:
The goal here is to do something quickly that gets our CI back to correctly running, and then we can refine it later. But if my characterization of the state of the packages is incorrect we can try to do something more clever now.

Reply:
DLA-Future is one of the packages that uses `CMAKE_HIP_COMPILER`/`enable_language(HIP)`: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dla-future/package.py. It's not visible explicitly in the package, but with HIP in the build environment CMake finds it and sets `CMAKE_HIP_COMPILER` correctly. pika used to set `CMAKE_CXX_COMPILER` to hipcc, but no longer does it and now also relies on `CMAKE_HIP_COMPILER` being set: https://github.com/spack/spack/blob/9e2558bd563af31355d83471f628a5d02656cdac/var/spack/repos/builtin/packages/pika/package.py#L210-L219.

I understand the need to get the CI working as soon as possible. If this goes in, how temporary of a solution do you think this will be? I can certainly patch this change out for myself if/when it gets merged, but that feels a bit backwards.

Reply:
I guess part of the issue here is that only `CachedCMakePackage`s have had the implicit `CMAKE_CXX_COMPILER=hipcc` so far and those are the packages that may have been relying on it inadvertently. Regular `CMakePackage`s have had to do it correctly, or have a local fix. Putting these conflicts into `ROCmPackage` makes it affect also the `CMakePackage`s that were doing it correctly in the first place.

Could the added conflicts live in `CachedCMakePackage` instead, and replace
```
        # We defined hipcc as top-level compiler for packages when +rocm.
        # This avoid problems coming from rocm flags being applied to another compiler.
        if "+rocm" in spec:
            entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
```
? It's still semantically the wrong place, but it's as localized as the previous fix was and won't affect regular `CMakePackage`s?

Reply:
@msimberg we're going to change how we approach this, thanks for bringing this up.

Reply:
Nice, thanks @becker33!